contracts/liquidity_pool.cairo

```diff
@@ -635,6 +661,124 @@ func _get_value_of_pool_position{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+    let retval_adj = retval + 1;
+    return (retval_adj,);
+}
+
+// func get_value_of_capital_in_lpool{
+//     syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+// } () -> (
+// ) {
+
+// }
+
+@view
+func get_all_poolinfo{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+} () -> (
+    poolinfo_len: felt,
+    poolinfo: PoolInfo*
+) {
+    alloc_locals;
+    let (lptoken_addrs_len: felt, lptoken_addrs: Address*) = get_all_lptoken_addresses();
+    // map address->poolinfo via get_poolinfo
+    let (res: PoolInfo*) = alloc();
+    map_address_to_poolinfo(lptoken_addrs, res, lptoken_addrs_len, 0);
+    //let (array : PoolInfo*) = alloc();
+    //let pool = Pool(0, 0, 0);
+    //let poolinfo = PoolInfo(pool, 0, 0, 0, 0);
+    //assert array[0] = poolinfo;
+    return (lptoken_addrs_len, res);
+    //return (lptoken_addrs_len, array);
+}
+
+func map_address_to_poolinfo{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr} (
+    lpt_addrs: Address*,
+    poolinfo: PoolInfo*,
+    array_len: felt,
+    curr_index: felt
+) -> () {
+    let val = get_poolinfo(lpt_addrs[curr_index]);
+    assert poolinfo[curr_index] = val;
+    if(array_len == curr_index + 1){
+        return ();
+    }
+    return map_address_to_poolinfo(lpt_addrs, poolinfo, array_len, curr_index + 1);
+}
+
+// Returns UserPoolInfo, which is the value of user's capital in pool and PoolInfo.
+func get_userpoolinfo{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+}(
+    user_address: Address,
+    lptoken_address: Address
+) -> UserPoolInfo {
+    alloc_locals;
+    let poolinfo = get_poolinfo(lptoken_address);
+    let (balance: Uint256) = ILPToken.balanceOf(contract_address=lptoken_address, account=user_address);
+    if (balance.low == 0 and balance.high == 0){
+        let nonzero_val = Uint256(1, 0);
+        let res = UserPoolInfo(nonzero_val, poolinfo);
+        return res;
+    }
+
+    let (currency_address) = get_underlying_token_address(lptoken_address); // This is unnecessarry, right?
+    let (total_lpt: Uint256) = ILPToken.totalSupply(contract_address=lptoken_address);
+    let total_value_of_pool_m64x61: Math64x61_ = Math64x61.add(poolinfo.unlocked_capital, poolinfo.value_of_pool);
+    let total_value_of_pool: Uint256 = toUint256(total_value_of_pool_m64x61, currency_address);
+    // User_value = (lpt_balance/total_lpt) * pool_balance = lpt_balance * pool_balance / total_lpt
+    let (a_low, a_high) = uint256_mul(balance, total_value_of_pool);  //uint256_unsigned_div_rem(balance, total_lpt);
+    let (b_div, b_rem) = uint256_unsigned_div_rem(a_low, total_lpt);
+    // returns less than it should in edge cases, but by the time we need this, Cairo 1.0 overhauls this and we are rewriting all the view functions anyway
+    let res = UserPoolInfo(b_div, poolinfo);
+    return res;
```

Comment:
where is the value of pool's position? why not to use `get_value_of_pool_position`

Reply:
or even better get_underlying_for_lptokens